streamz/tests/test_kafka.py

```diff
@@ -0,0 +1,96 @@
+import atexit
+from contextlib import contextmanager
+import os
+import pytest
+import random
+import requests
+import subprocess
+from time import sleep
+from tornado import gen
+
+from ..core import Stream
+from streamz.utils_test import gen_test
+
+j = random.randint(0, 10000)
+KAFKA_FILE = 'kafka_2.11-1.0.0'
+LAUNCH_KAFKA = os.environ.get('STREAMZ_LAUNCH_KAFKA', '') == 'true'
+ARGS = {'bootstrap.servers': 'localhost:9092', 'group.id': 'streamz-test%i' % j}
+TOPIC = 'test'
+ck = pytest.importorskip('confluent_kafka')
+
+
+def download_kafka(target):
+    r = requests.get('http://apache.mirror.globo.tech/kafka/1.0.0/'
+                     '%s.tgz' % KAFKA_FILE, stream=True)
+    with open(target, 'wb') as f:
+        for chunk in r.iter_content(2 ** 20):
+            f.write(chunk)
+    subprocess.check_call(['tar', 'xzf', KAFKA_FILE],
+                          cwd=os.path.dirname(target))
+
+
+def launch_kafka():
+    import shlex
+    cmd = shlex.split(
+        "docker run -d -p 2181:2181 -p 9092:9092 --env "
+        "ADVERTISED_HOST=127.0.01 --env ADVERTISED_PORT=9092 "
+        "--name streamz-kafka spotify/kafka")
+    cid = subprocess.check_output(cmd).decode()[:-1]
+
+    def end():
+        if cid:
+            subprocess.call(['docker', 'kill', cid])
+            subprocess.call(['docker', 'rm', cid])
+    atexit.register(end)
+
+    timeout = 10
+    while True:
+        try:
+            out = subprocess.check_output(['docker', 'logs', cid],
+                                          stderr=subprocess.STDOUT)
+            if b'kafka entered RUNNING state' in out:
+                print('Kafka is up')
+                break
+        except subprocess.CalledProcessError:
+            pass
+        sleep(0.1)
+        timeout -= 0.1
+        if timeout < 0:
+            raise RuntimeError('Timeout waiting for kafka')
+
+
+@contextmanager
+def kafka_service():
+    if LAUNCH_KAFKA:
+        launch_kafka()
+
+    producer = ck.Producer({'bootstrap.servers': 'localhost:9092'})
+    producer.produce('test', b'test')
+    out = producer.flush(10)
+    if out > 0:
+        raise RuntimeError('Timeout waiting for kafka')
+    yield producer
+
+
+@gen_test(timeout=60)
+def test_from_kafka():
+    with kafka_service() as kafka:
+        stream = Stream.from_kafka([TOPIC], ARGS)
+        yield gen.sleep(3)  # empty any lingering messages
+        out = stream.sink_to_list()
+        for i in range(10):
+            kafka.produce(TOPIC, b'value-%d' % i)
+        kafka.flush()
+        # it takes some time for messages to come back out of kafka
+        yield gen.sleep(2)
```

Comment:
Is there something we can wait on or poll here instead?  Fixed sleeps like this make me nervous.  I find that they often come up as causes of intermittent failures.  I often do something like the following:

```python
start = time()
while not condition:
    yield gen.sleep(0.1)
    assert time() < start + 5
```